src/lightning_app/core/app.py

```diff
@@ -521,6 +527,83 @@ def state_dict(self) -> t.Dict:
+        checkpoint_dict = None
+        if os.path.exists(checkpoint):
+            checkpoint_dict = self._load_checkpoint_from_json_file(checkpoint)
+        if os.path.exists(os.path.join(self.checkpoint_dir, checkpoint)):
+            checkpoint_dict = self._load_checkpoint_from_json_file(os.path.join(self.checkpoint_dir, checkpoint))
+
+        return checkpoint_dict
+
+    def _get_checkpoint_if_available_on_drive(self, checkpoint: str) -> t.Optional[dict]:
+        drive = Drive("lit://checkpoints", component_name="root")
+
+        with _context(ComponentContext.WORK):
+            found_checkpoints = drive.list(self.checkpoint_dir)
+            logger.debug("Found checkpoints: %s", found_checkpoints)
+
+        checkpoint_drive_path = ""
+
+        if checkpoint == "latest":
+
+            r = re.compile(LIGHTNING_CHECKPOINT_FILENAME_REGEX)
+            filtered_list = list(filter(r.search, found_checkpoints))
+            checkpoint_drive_path = max(filtered_list, key=lambda i: int(r.search(i).group(1))) if filtered_list else ""
+        else:
+            for cp in found_checkpoints:
+                if checkpoint in cp:
+                    checkpoint_drive_path = cp
+                    break
+
+        # Cannot file the checkpoint on the drive
+        if not checkpoint_drive_path:
+            return None
+
+        logger.debug("Downloading checkpoint from drive %s", checkpoint_drive_path)
+
+        # download the checkpoint from the drive
+        with _context(ComponentContext.WORK):
+            drive.get(checkpoint_drive_path)
+
+        if os.path.exists(checkpoint_drive_path):
+            checkpoint_dict = self._load_checkpoint_from_json_file(checkpoint_drive_path)
+        elif os.path.exists(os.path.join(self.checkpoint_dir, checkpoint_drive_path)):
+            checkpoint_dict = self._load_checkpoint_from_json_file(
+                os.path.join(self.checkpoint_dir, checkpoint_drive_path)
+            )
+        else:
+            raise FileNotFoundError(
+                f"Could not read checkpoint file: {checkpoint_drive_path} although it was found on the drive."
+            )
+
+        return checkpoint_dict
+
+    def load_app_state_from_checkpoint(self, checkpoint: str) -> None:
+        """Load the app state from a checkpoint.
+
+        Args:
+            checkpoint: The checkpoint to load the state from. Can be either:
+                - a path to a checkpoint file
+                - checkpoint file name in the .checkpoints dir in the drive
+                - 'latest' to load the latest saved checkpoint.
+        """
+        logger.info("Loading app state from checkpoint %s", checkpoint)
+
+        state = self._get_checkpoint_if_available_locally(checkpoint)
+        if not state:
+            state = self._get_checkpoint_if_available_on_drive(checkpoint)
+
+        if not state:
+            raise FileNotFoundError(f"Could not find checkpoint: {checkpoint}")
+        self.load_state_dict(state)
```

Comment:
This won't work with dynamic components. We need to finish the logic with load_state and get_state.

Reply:
Yes, supporting dynamic components might be tricky. Let me try to fix it, otherwise I will add it in another PR.

Reply:
One challenge with dynamic works is how to load the component and initialize it from the saved checkpoint, do you have ideas to do it, maybe pickle ? @tchaton

Reply:
Or it should be expected from the user to implement `load_state_dict()` and Re-instantiate the dynamic components or have it in a separate hook.

Reply:
updated